crates/network/src/lib.rs

```diff
@@ -0,0 +1,392 @@
+                                    server_port: p.server_port,
+                                    next_state: p.next_state,
+                                });
+                                connection.state = p.next_state;
+                            }
+                        },
+                        Err(e) => {
+                            error!("error: {:?}", e)
+                        }
+                    };
+                }
+                HandshakeState::Status => {
+                    match ClientStatusPacket::decode(&mut std::io::Cursor::new(payload)) {
+                        Ok(p) => match p {
+                            ClientStatusPacket::Request(_) => {
+                                client_status_request_events.send(ClientStatusRequestEvent {
+                                    connection: packet.connection,
+                                })
+                            }
+                            ClientStatusPacket::Ping(ping) => {
+                                connection.send_packet(
+                                        rjacraft_protocol::packets::server::ServerStatusPacket::Pong(
+                                            rjacraft_protocol::packets::server::Pong {
+                                                payload: ping.payload
+                                            }
+                                        )
+                                    ).unwrap();
+                                connection.flush_packets();
+                            }
+                        },
+                        Err(e) => {
+                            error!("error: {:?}", e)
+                        }
+                    };
+                }
+                HandshakeState::Login => {}
+                HandshakeState::Configuration => {}
+                HandshakeState::Play => {}
+            }
+        }
+    }
+}
+
+#[derive(Resource, Clone)]
+pub struct SharedNetworkState(Arc<SharedNetworkStateInner>);
+
+struct SharedNetworkStateInner {
+    new_connections_send: Sender<RemoteConnection>,
+    new_connections_recv: Receiver<RemoteConnection>,
+}
+
+#[derive(Component)]
+pub struct RemoteConnection {
+    pub remote_addr: SocketAddr,
+    recv: Receiver<PacketFrame>,
+    send: Sender<BytesMut>,
+    encoder: PacketEncoder,
+    recv_task: JoinHandle<()>,
+    send_task: JoinHandle<()>,
+    pub state: HandshakeState,
+}
+
+impl RemoteConnection {
+    pub fn try_recv(&mut self) -> Result<PacketFrame, TryRecvError> {
+        self.recv.try_recv()
+    }
+
+    pub fn write_packet_bytes(&mut self, bytes: &[u8]) -> anyhow::Result<()> {
+        self.encoder.append_frame(bytes)
+    }
+
+    pub fn flush_packets(&mut self) -> Result<(), TrySendError<BytesMut>> {
+        let bytes = self.encoder.take();
+        if !bytes.is_empty() {
+            self.send.try_send(bytes)
+        } else {
+            Ok(())
+        }
```

Comment:
I would prefer the branches reverted for better readability:

```suggestion
        if bytes.is_empty() {
            Ok(())
        } else {
            self.send.try_send(bytes)
        }
```